### PR TITLE
refactored the leaderboard controller to query from leaderboard model

### PR DESCRIPTION
**Pull request by DiptoSarkar182**
Before it was six different routes for fetching leader-boards based on daily, weekly and monthly. Now it is using two routes and those timeline are passed  as param value.

### CODE REVIEW

File: app/controllers/friend_lists_controller.rb [hunk 1]
```diff
@@ -3,8 +3,7 @@ class FriendListsController < ApplicationController
   before_action :check_token_expiration
 
   def index
-    @current_user_friend_lists = current_user.friends
-                                             .select("friend_lists.id AS id, friend_lists.user_id, users.id AS friend_id, users.full_name, users.level")
+    @current_user_friend_lists = FriendList.current_user_friend_lists(current_user)
```
Comment (JakariaJishan) [inline app/controllers/friend_lists_controller.rb:6]: If you access the current user globally, then there is no need to pass it as a parameter. You can use it at model level

File: app/controllers/friend_lists_controller.rb [hunk 2]
```diff
@@ -14,44 +13,14 @@ def index
   end
 
   def add_friend
-    friend_id = params[:friend_id]
+    result = FriendList.add_friend_in_friend_lists(current_user, params[:friend_id])
```
Comment (JakariaJishan) [inline app/controllers/friend_lists_controller.rb:16]: same thing goes here

File: app/controllers/friend_requests_controller.rb [hunk 3]
```diff
@@ -4,13 +4,7 @@ class FriendRequestsController < ApplicationController
 
 
   def current_user_sent_friend_requests
```
Comment (JakariaJishan) [inline app/controllers/friend_requests_controller.rb:6]: I think it's better to rename this method as send_friend_requests

File: app/controllers/friend_requests_controller.rb [hunk 4]
```diff
@@ -20,13 +14,7 @@ def current_user_sent_friend_requests
   end
 
   def current_user_received_friend_requests
```
Comment (JakariaJishan) [inline app/controllers/friend_requests_controller.rb:16]: same goes here

File: app/controllers/friend_requests_controller.rb [hunk 5]
```diff
@@ -36,84 +24,43 @@ def current_user_received_friend_requests
   end
 
   def send_friend_request
```
Comment (JakariaJishan) [inline app/controllers/friend_requests_controller.rb:26]: I am confused about current user send friend request and send friend request

File: app/controllers/leaderboards_controller.rb [hunk 6]
```diff
@@ -2,105 +2,39 @@ class LeaderboardsController < ApplicationController
...
+      1.day.ago
+    when "weekly"
+      7.days.ago
+    when "monthly"
+      30.days.ago
     else
-      render json: { message: 'No weekly leaderboard available' }, status: :ok
+      30.days.ago
     end
   end
```
Comment (JakariaJishan) [inline app/controllers/leaderboards_controller.rb]: I think its better if you declare this method to model level.

File: app/models/friend_list.rb [hunk 7]
```diff
@@ -1,5 +1,59 @@
...
+        .select("friend_lists.id AS id, friend_lists.user_id, users.id AS friend_id, users.full_name, users.level")
+        .order("users.full_name")
+  end
+
+  def self.add_friend_in_friend_lists(current_user, friend_id)
+    friend = User.find_by(id: friend_id)
+
+    if friend.nil?
+      return { status: :not_found, message: "The user you are trying to add does not exist" }
+    end
```
Comment (JakariaJishan) [inline app/models/friend_list.rb:17]: Is there any possibility of being a friend empty? Ultimately you are sending a friend request to a user. If there are no users you can not send friend reqeust.

File: app/models/friend_list.rb [hunk 8]
```diff
@@ -1,5 +1,59 @@
...
+
+    if current_user.id == friend_id
+      return { status: :unprocessable_entity, message: "You can't add yourself as a friend" }
+    end
+
+    # Check if a friend request exists where the current user is the receiver
+    request = FriendRequest.find_by(sender_id: friend_id, receiver_id: current_user.id)
+    if request.nil?
+      return { status: :unprocessable_entity, message: "You need to receive a friend request from this user to add them" }
+    end
```
Comment (JakariaJishan) [inline app/models/friend_list.rb:27]: Maybe It's better if you define the method in model level

File: app/models/friend_list.rb [hunk 9]
```diff
@@ -1,5 +1,59 @@
...
+
+    # Proceed to add to the friend list since the request exists and current_user is accepting it
+    ActiveRecord::Base.transaction do
+      # Add current_user and friend_id as friends
+      current_user.friend_lists.create!(friend_id: friend_id)
+      friend.friend_lists.create!(friend_id: current_user.id)
+
+      # Optionally, destroy the friend request after accepting
+      request.destroy
+    end
```
Comment (JakariaJishan) [inline app/models/friend_list.rb:37]: Good approach of using transactions

File: app/models/friend_request.rb [hunk 10]
```diff
@@ -2,4 +2,97 @@ class FriendRequest < ApplicationRecord
...
+
+  def self.current_user_received_friend_requests(user)
+    user.received_friend_requests.map do |request|
+      {
+        request_id: request.id,
+        sender_id: request.sender.id,
+        full_name: request.sender.full_name
+      }
+    end
+  end
```
Comment (JakariaJishan) [inline app/models/friend_request.rb]: You are preparing a response, right? I think it's better to use serializer

File: app/models/friend_request.rb [hunk 11]
```diff
@@ -2,4 +2,97 @@ class FriendRequest < ApplicationRecord
...
+      return { status: :unprocessable_entity, message: "You have already received a friend request from this user" }
+    end
+
+    friend_request = sender.sent_friend_requests.new(receiver: receiver)
+    if friend_request.save
+      { status: :ok, message: "Friend request sent successfully" }
+    else
+      { status: :unprocessable_entity, message: "Failed to send friend request", errors: friend_request.errors.full_messages }
+    end
+  end
```
Comment (JakariaJishan) [inline app/models/friend_request.rb]: The method is too large. Can you reduce this method by removing unnecessary checking?

File: app/models/friend_request.rb [hunk 12]
```diff
@@ -2,4 +2,97 @@ class FriendRequest < ApplicationRecord
...
+    if friend_request.nil?
+      return { status: :not_found, message: "No friend request found to cancel" }
+    end
+
+    if friend_request.destroy
+      { status: :ok, message: "Friend request canceled successfully" }
+    else
+      { status: :unprocessable_entity, message: "Failed to cancel the friend request", errors: friend_request.errors.full_messages }
+    end
+  end
```
Comment (JakariaJishan) [inline app/models/friend_request.rb]: same things here. so many checkings. please refactor this

File: app/models/leaderboard.rb [hunk 13]
```diff
@@ -2,4 +2,22 @@ class Leaderboard < ApplicationRecord
...
+
+  def self.leaderboards_excluding_top_three(time_range)
+    top_three_user_ids = top_three_users(time_range).pluck("users.id")
+
+    joins(:user)
+      .select("leaderboards.id, users.id AS user_id, users.full_name, leaderboards.points, leaderboards.created_at, leaderboards.updated_at")
+      .where("leaderboards.updated_at >= ?", time_range)
+      .where.not(user_id: top_three_user_ids)
+      .order(points: :desc)
+  end
```
Comment (JakariaJishan) [inline app/models/leaderboard.rb]: Good approach, Appriciated.

**Review by JakariaJishan — CHANGES_REQUESTED** [review]
Please address all of the comments.

File: app/controllers/friend_requests_controller.rb (hunk 3 above)
Comment (DiptoSarkar182) [reply app/controllers/friend_requests_controller.rb:6]: current_user_sent_friend_requests (GET request) here means that the request I sent to other users.

File: app/controllers/friend_requests_controller.rb (hunk 4 above)
Comment (DiptoSarkar182) [reply app/controllers/friend_requests_controller.rb:16]: current_user_received_friend_requests (GET request) here means that the request I received from other users.

File: app/controllers/friend_requests_controller.rb (hunk 5 above)
Comment (DiptoSarkar182) [reply app/controllers/friend_requests_controller.rb:26]: This actualy sends friend request to other user.

File: app/models/friend_list.rb (hunk 7 above)
Comment (DiptoSarkar182) [reply app/models/friend_list.rb:17]: The reason I added this check is to account for a situation where a user sends a friend request, and the receiver sees it in their request panel but, before accepting, the sender deletes their account.

File: app/controllers/leaderboards_controller.rb (hunk 6 above)
Comment (DiptoSarkar182) [reply app/controllers/leaderboards_controller.rb]: Moved this to leaderboard model.

File: app/models/friend_request.rb (hunk 10 above)
Comment (DiptoSarkar182) [reply app/models/friend_request.rb]: Added serializer.

File: app/models/friend_request.rb (hunk 11 above)
Comment (DiptoSarkar182) [reply app/models/friend_request.rb]: Moved the checkings to private class method.

File: app/models/friend_request.rb (hunk 12 above)
Comment (DiptoSarkar182) [reply app/models/friend_request.rb]: Moved the checkings to private class method.

File: app/models/sub_category_follower.rb [hunk 14]
```diff
@@ -2,4 +2,40 @@ class SubCategoryFollower < ApplicationRecord
...
+
+    return { message: "You are not following this subcategory", status: :unprocessable_entity } if sub_category_follower.nil?
+
+    if sub_category_follower.destroy
+      sub_category.decrement!(:total_follower)
+      { status: :ok, message: "Successfully unfollowed the subcategory", sub_category: sub_category }
+    else
+      { message: "Failed to unfollow the subcategory", errors: sub_category_follower.errors.full_messages, status: :unprocessable_entity }
+    end
+  end
```
Comment (JakariaJishan) [inline app/models/sub_category_follower.rb:40]: If possible can you refactor this? Looks like redundant code.

**Review by JakariaJishan — APPROVED** [review]
Everything looks good. 
Congrats. 
Approved.

**Pull request merged**